sdk/core/core-rest-pipeline/src/nodeHttpClient.ts

```diff
@@ -243,28 +246,55 @@ class NodeHttpClient implements HttpClient {
   }
 
   private getOrCreateAgent(request: PipelineRequest, isInsecure: boolean): http.Agent {
-    if (!request.disableKeepAlive) {
-      if (isInsecure) {
-        if (!this.httpKeepAliveAgent) {
-          this.httpKeepAliveAgent = new http.Agent({
-            keepAlive: true,
-          });
-        }
+    const disableKeepAlive = request.disableKeepAlive;
 
-        return this.httpKeepAliveAgent;
-      } else {
-        if (!this.httpsKeepAliveAgent) {
-          this.httpsKeepAliveAgent = new https.Agent({
-            keepAlive: true,
-          });
-        }
+    // Handle Insecure requests first
+    if (isInsecure) {
+      if (disableKeepAlive) {
+        // keepAlive:false is the default so we don't need a custom Agent
+        return http.globalAgent;
+      }
 
-        return this.httpsKeepAliveAgent;
+      // Get the cached agent if exists
+      let agent = this.cachedHttpAgent;
+      if (!agent) {
+        // If there is no cached agent create a new one and cache it.
+        agent = new http.Agent({ keepAlive: true });
+        this.cachedHttpAgent;
       }
-    } else if (isInsecure) {
-      return http.globalAgent;
+      return agent;
     } else {
-      return https.globalAgent;
+      // We use the tlsSettings to index cached clients
+      const tlsSettings = request.tlsSettings ?? DEFAULT_TLS_SETTINGS;
+      if (disableKeepAlive && !request.tlsSettings) {
+        // When there are no tlsSettings and keepAlive is false
+        // we don't need a custom agent
+        return https.globalAgent;
+      }
+
+      // Get the cached agent or create a new one with the
+      // provided values for keepAlive and tlsSettings
+      let agent =
+        this.cachedHttpsAgents.get(tlsSettings) ??
+        new https.Agent({
+          // keepAlive is true if disableKeepAlive is false.
+          keepAlive: !disableKeepAlive,
+          // Since we are spreading, if non tslSettings were provided, nothing is added to the agent options.
+          ...tlsSettings,
+        });
+
+      // If there was a cached client, check if we can return it.
+      if (this.cachedHttpsAgents.has(tlsSettings)) {
+        // We have a cached agent, check if keepAlive matches with the cache
+        if (agent.options.keepAlive === !disableKeepAlive) {
```

Comment:
can we combine these two if clauses?

Reply:
I think I can make this else clause simpler

Reply:
Let me know what you think about the refactored way